doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,258 @@
+   +----+------------------+
+
+             legend
+             +----------------+
+             |cGRE OSD stages |
+             +----------------+
+             |cBLU PG stage   |
+             +----------------+
+
+Two Requirements
+----------------
+
+An OSD client talks with each connected OSD over stateful connection. The client is allowed to
+perform asynchronous I/O with OSD in the sense that it can send another request without waiting
+for the completion of the previous one. But the contract of the RADOS protocol requires writes
+(and write ordered reads) performed on a single object to be performed in the order in which
+they were originally submitted, and only once. This behavior is implemented by both the client
+side and server side of librados.
+
+So far, we have two restrictions:
+
+#. sequential stages of a given request
+#. ordered execution of multiple write requests on a single object
+
+It's like visiting a history museum with friends.
+
+#. since the exhibition rooms in the museum are arranged in chronological order, it is advisable
+   to visit these rooms in the same order.
+#. before leaving the museum, visitors can leave a message on one of the notebooks provided by
+   the museum. But the visitors should do this in the same order in which they enter the museum.
+   For example, if both Joe and Jane decide to leave a message on the notebook with purple cover,
+   and Joe enters the museum before Jane, Jane cannot write on the notebook until Joe does.
+
+Current Implementation
+======================
+
+We, as the staff of the museum, are responsible to codify these guidelines. It is straightforward
+to follow the first rule. We can setup a pipeline for the exhibition rooms so each room will be
+a stage in the pipeline. Once a visitor finishes visiting a room, he or she will be transferred to
+the next room until the last one. But regarding the second rule, we have a couple solutions:
+
+- prohibit more than 1 person staying in a single room. The notebooks are located in the last room.
+  This way, the order in which visitors leave messages would be exactly the same as the order they
+  enter the museum.
+- allow more than 1 person staying in a single room as long as the number of the visitor does not
+  exceeds the capacity of the room. The staff of museum should give the purple handbook to the first
+  visitor who registers to write on this handbook, so the visitor can hand it over to the next one
+  or simply return it to the staff once she leaves her message on it. Visitors can pick a notebook
+  before entering the museum, and they should follow this protocol.
+
+Apparently, the first solution is way simpler. It **satisfies** both of the requirements, but its
+downside is that a slow visitor could keep others out of the door for long time, and hence increases
+the overall waiting time. This could be worse if we have multiple visitors who keep the same visitor(s)
+waiting for different rooms.
```

Comment:
I'm not sure this is a good analogy.  Most of the pipeline stages have the property that once someone passes through it, every subsequent request with the same requirement will pass through immediately (once we have OSDMap 20, every other message requiring OSDMap 20 will pass through instantly).  Also, if some request stops (let's say at wait for object context), every subsequent request would also have to block as they'd all need the object context with locks for that object.  For those cases, it's truly irrelevant whether one request or several are "in the room" -- once one request has requested the resource, we're making as much progress as we can.

(@tchaikov This paragraph may be particularly interesting.)  The get_obc PG pipeline stage is somewhat special.  While in the get_obc "room", we perform a read to obtain the object context for the object in question (which can block), and we obtain locks on the obc required to perform the IO.  Because reads *must* not see uncommitted write state, we cannot let a read through until all pending writes on the object have completed.  Similarly, we can't permit a write through until pending reads complete -- it's a bit less obvious, but if the previously submitted read has for whatever reason not completed yet, the uncommitted state from the write might appear before the read IO resulting in the same read of uncommitted state.  Thus, an op may have to block here until previous IOs complete.  Suboptimally, we do this *across objects*.  If we have read requests on foo, bar, and baz all in the same pg and foo is up next but blocked on a previous write completing, we could easily choose to let bar through first, but the current pipeline won't allow it because the bar request will be stuck "outside of the room".  This would be a pretty straightforward optimization, but it's not really a matter of allowing multiple people in the room so much as it is a matter of directly integrating the rwstate logic with the pipeline stages.

The process state is the other interesting case.  At this point, any pure write won't really block as it already has the locks/state it needs (ObjectContext generally) and will simply assemble an in-memory transaction and submit it to the distributed state machine.  Any read-modify-write would actually be required to block until the previous writes complete (or the read will be wrong) before doing its reads for the same reason that a read must.

(@tchaikov also probably interesting) In the case of reads, strictly speaking, we could permit multiple reads to occur here even with concurrent writes.  The get_obc stage above would already have ensured that any particular object can only have reads or writes here, so it's not a correctness problem.  But the pg state itself would still be being mutated by writes on other objects, so PG state variables wouldn't necessarily be stable for the duration of handling any particular read.  When you consider that the OSD (device) as a whole will have 100 or more PGs and a particular partition would likely still have 10 or more, it seems unlikely to be particularly useful as a way of increasing read concurrency, at least not yet.

Reply:
I am still digesting this comment..